### PR TITLE
[Shop][Payment] Prevent selecting payment method not assigned to channel

### PR DESCRIPTION
**Pull request by tuka217**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Related tickets | -
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentMethodRepository.php [hunk 1]
```diff
@@ -26,4 +28,18 @@ public function createListQueryBuilder()
             ->leftJoin('o.translations', 'translation')
         ;
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findEnabledForChannel(ChannelInterface $channel)
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentMethodRepository.php]: Why won't we just use `$channel->getPaymentMethods()` collection? It implements `Doctrine\Common\Collections\Selectable`, so:

```php
$channel->getPaymentsMethods()->matching(new Criteria(new Comparison('enabled', Comparison::EQ, true))
```

Should work even faster (we won't need to hydrate all the payment methods only to pass them to another query).

Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentMethodRepository.php]: Then we would not need this custom repository, we can use it directly in the resolver or implement on the model - see http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/working-with-associations.html#filtering-collections.

File: src/Sylius/Component/Core/Resolver/ChannelBasedPaymentMethodsResolver.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $this->paymentMethodRepository = $paymentMethodRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSupportedMethods(BasePaymentInterface $payment)
+    {
+        Assert::true($this->supports($payment));
```
Comment (pamil) [inline src/Sylius/Component/Core/Resolver/ChannelBasedPaymentMethodsResolver.php]: Custom message may be nicer than "expected true got false" :)

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentMethodRepository.php (hunk 1 above)
Comment (tuka217) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentMethodRepository.php]: Method 'mathing' is allowed only for `ArrayCollection`. We have here an instance of `Collection` class.

Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentMethodRepository.php]: Of course the typehint is oversimplified now, it's either `ArrayCollection` (if created manually) or `PersistentCollection` (if hydrated by Doctrine), but both of them implements `Collection` and `Selectable`.

Comment (tuka217) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentMethodRepository.php]: Yups :)

File: src/Sylius/Component/Core/Repository/PaymentMethodRepositoryInterface.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Repository;
+
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Sylius\Component\Payment\Repository\PaymentMethodRepositoryInterface as BasePaymentMethodRepositoryInterface;
+/**
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Repository/PaymentMethodRepositoryInterface.php]: Missing blank line

File: src/Sylius/Component/Core/Resolver/ChannelBasedPaymentMethodsResolver.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Repository\PaymentMethodRepositoryInterface;
+use Sylius\Component\Payment\Model\PaymentInterface as BasePaymentInterface;
+use Sylius\Component\Payment\Resolver\MethodsResolverInterface;
+use Webmozart\Assert\Assert;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+class ChannelBasedPaymentMethodsResolver implements MethodsResolverInterface
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Resolver/ChannelBasedPaymentMethodsResolver.php]: final

File: src/Sylius/Component/Core/Resolver/ChannelBasedPaymentMethodsResolver.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+{
+    /**
+     * @var PaymentMethodRepositoryInterface
+     */
+    protected $paymentMethodRepository;
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Resolver/ChannelBasedPaymentMethodsResolver.php]: private

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentMethodRepository.php (hunk 1 above)
Comment (tuka217) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentMethodRepository.php]: In current implementation it cannot be done in this way.

**Comment by pjedrzejewski** [comment]
Thank you Ania! :)

**Pull request merged**